---
params:
  study: "Study Name"
  metadata_path: "path/to/metadata.csv"
  bismark_path: "path/to/bismark"
  gtf_path: "path/to/gtf"
  chromhmm_path: "path/to/chromhmm_segments.bed"
  ccre_path: "path/to/cCREs.bed"
  blacklist_path: "path/to/blacklist.bed"
  out_path: "path/to/methylkit"
  assembly: "genome_assembly"
title: |
  | Differential methylation analysis report
  | `r params$study`
date: "`r Sys.Date()`"
author: "`r Sys.info()[['user']]`"
version: "1.0"
output:
  rmdformats::readthedown:
    toc_depth: 6
    self_contained: yes
    
css: methylkit.css
---
<!--
# Differential Methylation Analysis Report
This R Markdown document performs differential methylation analysis for a specified study. 
It includes QC analysis, statistical analysis, and visualization of results.


Usage from R console:
rmarkdown::render(
  "bin/methylkit.rmd", 
  params = list(
    study = "JIRA_BDS-XXXX_12ABIC_EPI321_vs_CONTROL", 
    metadata_path = "/home/tylerborrman/epic-methylseq_data/test_12ABIC/samplesheet_test_12ABIC.csv",
    bismark_path = "/home/tylerborrman/epic-methylseq_data/test_12ABIC",
    gtf_path = "/home/tylerborrman/references/hg38/gtf/Homo_sapiens.GRCh38.109.cas_molecule.gtf",
    chromhmm_path = "/home/tylerborrman/epic-methylseq_data/ChromHMM/hg38_genome_100_segments.bed",
    ccre_path = "/home/tylerborrman/epic-methylseq_data/cCRE/GRCh38-cCREs.bed",
    blacklist_path = "/home/tylerborrman/epic-methylseq_data/blacklist/hg38-blacklist.v2.bed",
    out_path = "/home/tylerborrman/epic-methylseq_data/test_12ABIC/methylkit",
    assembly = "hg38"
  )
)

Usage from command line:
Rscript -e "rmarkdown::render(
  'bin/methylkit.rmd',
  params = list(
    study = 'JIRA_BDS-XXXX_12ABIC_EPI321_vs_CONTROL',
    metadata_path = '/home/tylerborrman/epic-methylseq_data/test_12ABIC/samplesheet_test_12ABIC.csv',
    bismark_path = '/home/tylerborrman/epic-methylseq_data/test_12ABIC',
    gtf_path = '/home/tylerborrman/references/hg38/gtf/Homo_sapiens.GRCh38.109.cas_molecule.gtf',
    chromhmm_path = '/home/tylerborrman/epic-methylseq_data/ChromHMM/hg38_genome_100_segments.bed',
    ccre_path = '/home/tylerborrman/epic-methylseq_data/cCRE/GRCh38-cCREs.bed',
    blacklist_path = '/home/tylerborrman/epic-methylseq_data/blacklist/hg38-blacklist.v2.bed',
    out_path = '/home/tylerborrman/epic-methylseq_data/test_12ABIC/methylkit',
    assembly = 'hg38'
  )
)"

params:
  study: Study Name with JIRA ID prefix
  metadata_path: path to samplesheet used in nfcore/methylseq pipeline with additional condition column
  bismark_path: path to directory containing bismark coverage files (.deduplicated.bismark.cov)
  gtf_path: path to GTF file for gene annotation
  chromhmm_path: path to ChromHMM annotation file
  ccre_path: path to ENCDOE cCRE annotation file
  blacklist_path = path to ENCODE blacklist annotation file
  out_path: path to directory to save methylkit output files
  assembly: genome assembly used in the analysis (e.g. hg38, mm10)
-->

```{r setup, include=FALSE} 
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache=FALSE)
# load packages
library(DT)
library(methylKit)
library(plotly)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(GenomicRanges)
library(rtracklayer)
library(pheatmap)
library(RColorBrewer)
library(MASS)
library(viridis)

# Specify number of cores to use
cores <- 80
```

```{r read_metadata}
# Load metadata
metadata <- read_csv(params$metadata_path)
metadata <- metadata[c("sample", "condition")]
metadata <- as.data.frame(metadata)
rownames(metadata) <- metadata$sample
```

```{r read_cov_files}

# Function to find files with a given prefix
find_files_with_prefix <- function(prefix, files) {
  matching_files <- grep(prefix, files, value = TRUE)
  if (length(matching_files) == 1) {
    return(matching_files)
  } else {
    stop(paste0("ERROR: No file or multiple files found for sample with prefix ", prefix))
  }
}
unordered_files <- list.files(
  params$bismark_path,
  pattern = ".cov",
  full.names = TRUE
)
# Ensure files are ordered according to metadata sample sheet
cov_files <- as.list(sapply(metadata$sample, find_files_with_prefix, files = unordered_files))

# Create variables for treated and control samples
if ("EPI321" %in% metadata$condition & "CONTROL" %in% metadata$condition) {
  treatment <- ifelse(metadata$condition == "EPI321", 1, 0)
  treated <- metadata$sample[metadata$condition == "EPI321"]
  control <- metadata$sample[metadata$condition == "CONTROL"]
  contrast <- c("EPI321", "CONTROL")
} else if ("6-MONTH" %in% metadata$condition & "3-MONTH" %in% metadata$condition) {
  treatment <- ifelse(metadata$condition == "6-MONTH", 1, 0)
  treated <- metadata$sample[metadata$condition == "6-MONTH"]
  control <- metadata$sample[metadata$condition == "3-MONTH"]
  contrast <- c("6-MONTH", "3-MONTH")
} else if ("UD" %in% metadata$condition & "D2" %in% metadata$condition) {
  treatment <- ifelse(metadata$condition == "D2", 1, 0)
  treated <- metadata$sample[metadata$condition == "D2"]
  control <- metadata$sample[metadata$condition == "UD"]
  contrast <- c("D2", "UD")
} else if ("UD" %in% metadata$condition & "D7" %in% metadata$condition) {
  treatment <- ifelse(metadata$condition == "D7", 1, 0)
  treated <- metadata$sample[metadata$condition == "D7"]
  control <- metadata$sample[metadata$condition == "UD"]
  contrast <- c("D7", "UD")
} else if ("D2" %in% metadata$condition & "D7" %in% metadata$condition) {
  treatment <- ifelse(metadata$condition == "D7", 1, 0)
  treated <- metadata$sample[metadata$condition == "D7"]
  control <- metadata$sample[metadata$condition == "D2"]
  contrast <- c("D7", "D2")
} else {
  stop("Unknown contrast specified in samplesheet condition column")
}

# read methylation call files and store them as flat file database
# Note gzipped files are supported
# Note that the doc for methylKit says bismark coverage files have 5 cols, 
# but the bismark coverage files have 6 cols and the code for the methylKit methRead function expects 6 cols:
# chr, start, end, % meth, count methylated, count unmethylated
myobj=methRead(cov_files,
                sample.id=as.list(metadata$sample),
                pipeline="bismarkCoverage",
                assembly=params$assembly,
                treatment=treatment,
                context="CpG",
                header = FALSE,
                mincov = 3
)
# Merging all samples into one file. This is needed as precursor for dmc analyses  
meth=methylKit::unite(myobj, destrand=FALSE, mc.cores=cores)
```

```{r tiling_window_analysis, echo=FALSE}

# converting my.obj to tile object for sliding 100 bp windows minimum coverage of 5 CpGs 
invisible(capture.output({tile.obj = tileMethylCounts(myobj,win.size=100,step.size=100,cov.bases=5, mc.cores=cores)}))

# in this version of the code we only required 2 out of the 3 total merged samples (one sample per patient source) to contain a methylation site for it to be included in the comparative analyses if min.per.group is not specified default is all samples must have a specific CpG site or meth tile covered to include it. 
# meth.tile=methylKit::unite(tile.obj, destrand=FALSE, min.per.group = 2L) 

# In this version of the code we are requiring all samples to have a methylation site for it to be included in the comparative analyses.
meth.tile=methylKit::unite(tile.obj, destrand=FALSE, mc.cores=cores)
```

# Dataset overview

```{r dataset_overview}

hg38_names <- c(
  "hg38",
  "GRCh38",
  "Homo_sapiens.GRCh38.dna.primary_assembly.109.cas_molecule", 
  "Homo_sapiens.GRCh38.dna.primary_assembly.109.cas_molecule.fa"
)
mm10_names <- c(
  "mm10",
  "GRCm38",
  "Mus_musculus.GRCm38.dna.primary_assembly.102.cas_molecule", 
  "Mus_musculus.GRCm38.dna.primary_assembly.102.cas_molecule.fa"
)
macfas5_names <- c(
  "macFas5",
  "Macaca_fascicularis_5.0",
  "Macaca_fascicularis.Macaca_fascicularis_5.0.dna.toplevel.102.cas_molecule", 
  "Macaca_fascicularis.Macaca_fascicularis_5.0.dna.toplevel.102.cas_molecule.fa"
)

# All CpG and candidate DMR (tile) counts were calculated using
# *.cas_molecule.fa reference genomes 

if (params$assembly %in% hg38_names) {
  genome <- "Homo_sapiens.GRCh38.dna.primary_assembly.109.cas_molecule.fa"
  genome_cpg_cnt <- 58974418
  genome_tile_cnt <- 3176105
} else if (params$assembly %in% mm10_names) {
  genome <- "Mus_musculus.GRCm38.dna.primary_assembly.102.cas_molecule.fa"
  genome_cpg_cnt <- 43816302
  genome_tile_cnt <- 1960510
} else if (params$assembly %in% macfas5_names) {
  genome <- "Macaca_fascicularis.Macaca_fascicularis_5.0.dna.toplevel.102.cas_molecule.fa"
  genome_cpg_cnt <- 58009056
  genome_tile_cnt <- 3164794
} else {
  stop("Unknown reference genome assembly")
}
```

## Reference genome overview
Reference genome: ``r genome``

```{r genome_table}

genome_table <- data.frame(
  c(
    "Total CpGs",
    "Total tiles"
  ),
  c(
    genome_cpg_cnt,
    genome_tile_cnt
  )
)

colnames(genome_table) <- c("key", "value")

datatable(
  genome_table,
  rownames=FALSE,
  colnames=NULL,
  caption=htmltools::tags$caption(
    style = 'caption-side: bottom;',
    "Tile = 100 bp genome window with at least 5 CpGs."
  ),
  options=list(
    searching=FALSE,
    paging=FALSE,
    lengthChange=FALSE
  )
) %>% formatRound(
  columns = "value",
  digits = 0
)

```

## Sample overview

```{r sample_table}
sample_cpg_cnts <- sapply(myobj, function(x) nrow(x))
sample_tile_cnts <- sapply(tile.obj, function(x) nrow(x))

percent_cpg <- (sample_cpg_cnts / genome_cpg_cnt) * 100
percent_tile <- (sample_tile_cnts / genome_tile_cnt) * 100

metadata <- cbind(
  metadata,
  sample_cpg_cnts,
  percent_cpg,
  sample_tile_cnts,
  percent_tile
)

round_cols <- c(
  "percent_cpg",
  "percent_tile"
)

int_cols <- c(
  "sample_cpg_cnts",
  "sample_tile_cnts"
)

datatable(
  metadata,
  rownames = FALSE,
  colnames = c(
    "sample",
    "condition",
    "3X CpG",
    "3X CpG %",
    "3X Tile",
    "3X Tile %"
  ),
  caption= htmltools::tags$caption(
    style = 'caption-side: bottom;',  
    "3X CpG = Number of CpGs with at least 3X coverage.", htmltools::tags$br(),  
    "3X CpG % = Percentage of CpGs with at least 3X coverage of total CpGs.", htmltools::tags$br(), 
    "3X Tile = Number of tiles with at least 5 CpGs each with at least 3X coverage.", htmltools::tags$br(), 
    "3X Tile % = Percentage of tiles with at least 5 CpGs each with at least 3X coverage of total tiles."
  )
) %>% formatStyle(
  columns="percent_cpg",
  backgroundColor = styleInterval(75, c('#ff4545', 'white'))
) %>% formatStyle(
  columns="percent_tile",
  backgroundColor = styleInterval(75, c('#ff4545', 'white'))
) %>% formatRound(
  columns = round_cols,
  digits = 2
) %>% formatRound(
  columns = int_cols,
  digits = 0
)
```

## Coverage Statistics
CpGs and tiles must meet coverage requirements in all samples to be included in analysis.

```{r coverage_stats, results="asis"}
unite_percent_cpg <- round((nrow(meth) / genome_cpg_cnt) * 100, 2)
unite_percent_tile <- round((nrow(meth_tile) / genome_tile_cnt) * 100, 2)

if (unite_percent_cpg >= 75) {
  cat("CpGs with at least 3X coverage in all samples:") 
  cat(format(nrow(meth), big.mark = ","))
  cat("\n\nPercentage of CpGs with at least 3X coverage in all samples of total CpGs:    ")
  cat(paste0(unite_percent_cpg, "%"))
} else {
  cat("CpGs with at least 3X coverage in all samples: ")
  cat(paste0("<span style='color: red;'>", format(nrow(meth), big.mark = ",") , "</span>"))
  cat("\n\nPercentage of CpGs with at least 3X coverage in all samples of total CpGs:    ")
  cat(paste0("<span style='color: red;'>", unite_percent_cpg , "%</span>"))
}
if (unite_percent_tile >= 75) {
  cat("\n\nTiles with at least 5 CpGs each with at least 3X coverage in all samples:    ") 
  cat(format(nrow(meth.tile), big.mark = ","))
  cat("\n\nPercentage of tiles with at least 5 CpGs each with at least 3X coverage in all samples of total tiles:    ")
  cat(paste0(unite_percent_tile, "%"))
} else {
  cat("\n\nTiles with at least 5 CpGs each with at least 3X coverage in all samples:    ") 
  cat(paste0("<span style='color: red;'>", format(nrow(meth.tile), big.mark = ",") , "</span>"))
  cat("\n\nPercentage of tiles with at least 5 CpGs each with at least 3X coverage in all samples of total tiles:    ")
  cat(paste0("<span style='color: red;'>", unite_percent_tile , "%</span>"))
}
```

# Differentially methylated bases or regions

```{r diff_meth_cpg}

## Finding differentially methylated CpGs 
myDiff=calculateDiffMeth(meth, mc.cores=cores)
# find DMRs 
myDiff.tile=calculateDiffMeth(meth.tile, mc.cores=cores)

# Extracting data from the methylDiff objects and coverting to dataframes
meth.dmc <- getData(myDiff) 
meth.dmr <- getData(myDiff.tile) 
meth.dmc.df <- as.data.frame(meth.dmc)
meth.dmr.df <- as.data.frame(meth.dmr)
```
