---
params:
  study: "Study Name"
  metadata_path: "path/to/metadata.csv"
  bismark_path: "path/to/bismark"
  out_path: "path/to/methylkit"
  assembly: "genome_assembly"
title: |
  | Differential methylation analysis report
  | `r params$study`
date: "`r Sys.Date()`"
author: "`r Sys.info()[['user']]`"
version: "1.0"
output:
  rmdformats::readthedown:
    toc_depth: 6
    self_contained: yes
    
css: methylkit.css
---
<!--
# Differential Methylation Analysis Report
This R Markdown document performs differential methylation analysis for a specified study. 
It includes QC analysis, statistical analysis, and visualization of results.


Usage from R console:
rmarkdown::render(
  "bin/methylkit.rmd", 
  params = list(
    study = "JIRA_BDS-XXXX_12ABIC_EPI321_vs_CONTROL", 
    metadata_path = "/home/tylerborrman/epic-methylseq_data/test_12ABIC/samplesheet_test_12ABIC.csv",
    bismark_path = "/home/tylerborrman/epic-methylseq_data/test_12ABIC",
    out_path = "/home/tylerborrman/epic-methylseq_data/test_12ABIC/methylkit",
    assembly = "hg38"
  )
)

Usage from command line:
Rscript -e "rmarkdown::render(
  'bin/methylkit.rmd',
  params = list(
    study = 'JIRA_BDS-XXXX_12ABIC_EPI321_vs_CONTROL',
    metadata_path = '/home/tylerborrman/epic-methylseq_data/test_12ABIC/samplesheet_test_12ABIC.csv',
    bismark_path = '/home/tylerborrman/epic-methylseq_data/test_12ABIC',
    out_path = '/home/tylerborrman/epic-methylseq_data/test_12ABIC/methylkit',
    assembly = 'hg38'
  )
)"

params:
  study: Study Name with JIRA ID prefix
  metadata_path: path to samplesheet used in nfcore/methylseq pipeline with additional condition column
  bismark_path: path to directory containing bismark coverage files (.deduplicated.bismark.cov)
  out_path: path to directory to save methylkit output files
  assembly: genome assembly used in the analysis (e.g. hg38, mm10)
-->

```{r setup, include=FALSE} 
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,cache=FALSE)
# load packages
library(DT)
library(methylKit)
library(plotly)
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
```

# Dataset overview

```{r dataset overview}
# Load metadata
metadata <- read_csv(params$metadata_path)
metadata <- metadata[c("sample", "condition")]
datatable(metadata)
metadata <- as.data.frame(metadata)
rownames(metadata) <- metadata$sample
```

# Descriptive statistics

```{r descriptive statistics}

# Function to find files with a given prefix
find_files_with_prefix <- function(prefix, files) {
  matching_files <- grep(prefix, files, value = TRUE)
  if (length(matching_files) == 1) {
    return(matching_files)
  } else {
    stop(paste0("ERROR: No file or multiple files found for sample with prefix ", prefix))
  }
}
unordered_files <- list.files(
  params$bismark_path,
  pattern = ".cov",
  full.names = TRUE
)
# Ensure files are ordered according to metadata sample sheet
cov_files <- as.list(sapply(metadata$sample, find_files_with_prefix, files = unordered_files))

# Currently only supporting EPI321 vs Control comparison
treatment <- ifelse(metadata$condition == "EPI321", 1, 0)
treated <- metadata$sample[metadata$condition == "EPI321"]
control <- metadata$sample[metadata$condition == "CONTROL"]

# read methylation call files and store them as flat file database
# Note gzipped files are supported
# Note that the doc for methylKit says bismark coverage files have 5 cols, 
# but the bismark coverage files have 6 cols and the code for the methylKit methRead function expects 6 cols:
# chr, start, end, % meth, count methylated, count unmethylated

myobj=methRead(cov_files,
                sample.id=as.list(metadata$sample),
                pipeline="bismarkCoverage",
                assembly=params$assembly,
                treatment=treatment,
                context="CpG",
                dbtype = "tabix",
                dbdir = "methylDB",
                header = FALSE
)

# Descriptive statistics on samples
getMethylationStats(myobj[[1]],plot=TRUE,both.strands=FALSE) # check methylation stats on first sample
getCoverageStats(myobj[[1]],plot=TRUE,both.strands=FALSE) # check coverage stats on first  sample 
```

# Comparative Analysis
## CpG analysis
```{r comparative analysis}
# Comparative analysis
#  Merging all samples into one file. This is needed as precursor for dmc analyses  
meth=methylKit::unite(myobj, destrand=FALSE)
clusterSamples(meth, dist="correlation", method="ward.D2", plot=TRUE)
PCASamples(meth, screeplot=TRUE)
PCASamples(meth, adj.lim=c(0.5,0.5))

```


## Tiling windows analysis
For some situations, it might be desirable to summarize methylation information over tiling windows rather than doing base-pair resolution analysis.
For example, here we tile the genome with windows of 100bp and step-size 100bp and summarize the methylation information on those tiles.
This returns a methylRawList object which can be fed into unite and calculateDiffMeth functions consecutively to get differentially methylated regions.
The tilling function adds up C and T counts from each covered cytosine and returns a total C and T count for each tile.

```{r tiling_window_analysis}

# converting my.obj to tile object for sliding 100 bp windows minimum coverage of 5 CpGs 
tile.obj=tileMethylCounts(myobj,win.size=100,step.size=100,cov.bases=5)

# in this version of the code we only required 2 out of the 3 total merged samples (one sample per patient source) to contain a methylation site for it to be included in the comparative analyses if min.per.group is not specified default is all samples must have a specific CpG site or meth tile covered to include it. 
# meth.tile=methylKit::unite(tile.obj, destrand=FALSE, min.per.group = 2L) 

# In this version of the code we are requiring all samples to have a methylation site for it to be included in the comparative analyses.
meth.tile=methylKit::unite(tile.obj, destrand=FALSE)

PCASamples(meth.tile)
# find DMRs 
myDiff.tile=calculateDiffMeth(meth.tile)
```

# DMR no cut-off Tiling hyper/hypo methylated regions per chromosome
** In this section we will visualize the distribution of hypo/hyper-methylated bases/regions per chromosome.

```{r hyper/hypo_methylation}
# diffMethPerChr(myDiff.tile,plot=TRUE,qvalue.cutoff=1, meth.cutoff=25)
diffMethPerChr(myDiff.tile,plot=TRUE,qvalue.cutoff=1, meth.cutoff=21)


```

# Differentially methylated bases or regions
The calculateDiffMeth() function is the main function to calculate differential methylation.
Depending on the sample size per each set it will either use Fisher’s exact or logistic regression to calculate P-values.
P-values will be adjusted to Q-values using SLIM method (Wang, Tuominen, and Tsai 2011).
Note: if we have replicates, the function will automatically use logistic regression.
You can force the calculateDiffMeth() function to use Fisher’s exact test if you pool the replicates when there is only test and control sample groups.

```{r diff_meth_cpg}

## Finding differentially methylated CpGs (differentially methylated regions is in the previous section)
myDiff=calculateDiffMeth(meth)
```


```{r making_bed_files}

# Extracting data from the methylDiff objects and writing to bed files

## for tiling
select.meth <- getData(myDiff.tile) #DMRs

## for CpG sites 
select.meth2 <- getData(myDiff) #DMCs

# write bedfile
meth.bed <- as.data.frame(select.meth) #dmr
meth.bed2 <- as.data.frame(select.meth2)#dmc
write.table(meth.bed, paste0(params$out_path, "/", params$study,".meth.tile.all.bed"), sep="\t", col.names=FALSE, row.names=FALSE, quote=FALSE)
write.table(meth.bed2, paste0(params$out_path, "/", params$study,".meth.cpg.all.bed"), sep="\t", col.names=FALSE, row.names=FALSE, quote=FALSE)
```


```{r percent_methylation_tile}

perc.meth.tile=percMethylation(meth.tile, rowids = TRUE)
perc.meth.tile <- as.data.frame(perc.meth.tile)
perc.meth.tile$avg_treat <- rowMeans(perc.meth.tile[, colnames(perc.meth.tile) %in% treated], na.rm = TRUE)
perc.meth.tile$avg_ctrl <- rowMeans(perc.meth.tile[, colnames(perc.meth.tile) %in% control], na.rm = TRUE)
perc.meth.tile$coord <- rownames(perc.meth.tile)
meth.bed$coord <- paste0(meth.bed$chr,".",meth.bed$start,".",meth.bed$end)
perc.meth.tile <- merge(perc.meth.tile, meth.bed, by="coord")

# This is the final MR file we want!!!! 
write.table(
  perc.meth.tile,
  file = paste0(params$out_path, "/", params$study,".perc.meth.tile.all.tsv"),
  sep="\t",
  col.names=TRUE,
  row.names=FALSE,
  quote=FALSE
)
# This is the final DMR file we want!!!!
perc.meth.tile.sig <- perc.meth.tile[perc.meth.tile$qvalue <=0.01 & abs(perc.meth.tile$meth.diff) > 10,]

write.table(
  perc.meth.tile.sig,
  file = paste0(params$out_path, "/", params$study,".perc.meth.tile.all.q0.01.methdiff.10.tsv"),
  sep="\t",
  col.names=TRUE,
  row.names=FALSE,
  quote=FALSE
)
```

## Percent Scatterplots for significant tiles all

```{r scatter_tile}
## plotting only tiles with q <-0.05
# perc.meth.tile.plot <- merge(perc.meth.tile, meth.promoters, by=c("chr","end"))
# DT::datatable(perc.meth.tile.plot, extensions = 'Buttons',
#   options = list(
#     dom = 'Bfrtip',
#     buttons = c('copy', 'csv', 'excel', 'pdf', 'print')), caption =  paste0("Percent methylation at significant tiles for ",project, " q < 0.05"))
scatter_plot <- plot_ly(data = perc.meth.tile.sig, x = ~avg_treat, y = ~avg_ctrl, text = ~coord, type = 'scatter', mode = 'markers', marker = list(size = 10))

# Customize layout
scatter_plot <- scatter_plot %>% layout(
  xaxis = list(title = "Average Treatment"),
  yaxis = list(title = "Average Control"),
  hovermode = "closest",
  title = list(text = paste0("Scatter Plot of percent methylation at significant tiles sites for q < 0.01 and abs meth diff > 10"), font = list(size = 10)), font = list(size = 11))


# Print the plot
scatter_plot
```


```{r percent_methylation_cpg}

perc.meth.cpg=percMethylation(meth, rowids = TRUE)
perc.meth.cpg <- as.data.frame(perc.meth.cpg)
perc.meth.cpg$avg_treat <- rowMeans(perc.meth.cpg[,colnames(perc.meth.cpg) %in% treated], na.rm = TRUE)
perc.meth.cpg$avg_ctrl <- rowMeans(perc.meth.cpg[, colnames(perc.meth.cpg) %in% control], na.rm = TRUE)
perc.meth.cpg$coord <- rownames(perc.meth.cpg)
meth.bed2$coord <- paste0(meth.bed2$chr,".",meth.bed2$start,".",meth.bed2$end)
perc.meth.cpg <- merge(perc.meth.cpg, meth.bed2, by="coord")

# This is the final CpG file we want!!!! 
write.table(
  perc.meth.cpg,
  file = paste0(params$out_path, "/", params$study,".perc.meth.cpg.all.tsv"),
  sep="\t",
  col.names=TRUE,
  row.names=FALSE,
  quote=FALSE
)

# This is the final DMC file we want!!!
perc.meth.cpg.sig <- perc.meth.cpg[perc.meth.cpg$qvalue <=0.01 & abs(perc.meth.cpg$meth.diff) > 10,]
write.table(
  perc.meth.cpg.sig,
  file = paste0(params$out_path, "/", params$study,".perc.meth.cpg.all.q0.01.methdiff.10.tsv"),
  sep="\t",
  col.names=TRUE,
  row.names=FALSE,
  quote=FALSE
)
```

## Percent Scatterplots for significant CpG sites all

```{r, scatter_cpg}

scatter_plot <- plot_ly(data = perc.meth.cpg.sig, x = ~avg_treat, y = ~avg_ctrl, text = ~coord, type = 'scatter', mode = 'markers', marker = list(size = 10))

# Customize layout
scatter_plot <- scatter_plot %>% layout(
  xaxis = list(title = "Average Treatment"),
  yaxis = list(title = "Average Control"),
  hovermode = "closest",
  title = list(text = paste0("Scatter Plot of percent methylation at significant CpG sites for q < 0.01 and abs meth diff > 10"), font = list(size = 11))
)
# Print the plot
scatter_plot
```